File: SourceCode/pythonLib/0729_My_Calendar_I.py
class MyCalendar:
    def __init__(self):
        from sortedcontainers import SortedList
        self.data = SortedList()

    def book(self, start: int, end: int) -> bool:
        idx = self.data.bisect_right(start)
        if idx == len(self.data) or idx % 2 == 0 and self.data[idx] >= end:
            self.data.add(start)
            self.data.add(end)
            return True
        return False
        # Your MyCalendar object will be instantiated and called as such:
        # obj = MyCalendar()
        # param_1 = obj.book(start,end)

File: SourceCode/STL/0729_My_Calendar_I.cpp
#include <bits/stdc++.h>
using namespace std;
// by using lower_bound function in STL
// prev function
class MyCalendar {
private:
  set<pair<int, int>> ans;
public:
  MyCalendar() {
      
  }
  
  bool book(int start, int end) {
    auto iter = ans.lower_bound({end, INT_MIN});
    if(iter == ans.begin() || prev(iter)->second <= start) {
      ans.insert({start, end});
      return true;
    }
    return false;
  }
};

/**
 * Your MyCalendar object will be instantiated and called as such:
 * MyCalendar* obj = new MyCalendar();
 * bool param_1 = obj->book(start,end);
 */